### PR TITLE
Update to Clang 10 and Visual Studio 2019 version 16.6p2

### PR DESCRIPTION
**Pull request by BillyONeal**
* re-clang-formats the tree to comply with new clang 10 clang-format
* updates our agents to F series VMs which are faster and cheaper for our build workloads
* defaults scale set to 0 VMs and lets the Azure Pipelines service control all resizing
* fix tests to pass with the new compilers
* adds antivirus exclusions for the build directory

Co-authored by Casey Carter and Curtis Bezault

### CODE REVIEW

**Comment by miscco** [comment]
So for me I assume that these are exclusively formatting changes in the other files and all functional changes between clang-9 / clang-10 will only come at a later stage?

File: stl/inc/functional [hunk 1]
```diff
@@ -1536,12 +1523,11 @@ public:
...
     template <class... _Unbound>
-        constexpr auto operator()(_Unbound&&... _Unbargs) //
+    constexpr auto operator()(_Unbound&&... _Unbargs) //
         & noexcept(noexcept( //
             _Call_front_binder(_Seq{}, _Mypair._Get_first(), _Mypair._Myval2, _STD forward<_Unbound>(_Unbargs)...))) //
-          -> decltype( //
-              _Call_front_binder(
-                  _Seq{}, _Mypair._Get_first(), _Mypair._Myval2, _STD forward<_Unbound>(_Unbargs)...)) { //
+        -> decltype( //
+            _Call_front_binder(_Seq{}, _Mypair._Get_first(), _Mypair._Myval2, _STD forward<_Unbound>(_Unbargs)...)) { //
```
Comment (StephanTLavavej) [inline stl/inc/functional]: Empty comments were being used here to improve clang-format's behavior. If you remove them, does clang-format 10 do something reasonable-looking now?

File: stl/inc/random [hunk 2]
```diff
@@ -311,36 +311,28 @@ _NODISCARD _Uint _Next_linear_congruential_value(_Uint _Prev) noexcept {
...
+        return static_cast<_Uint>(_Cx); // relies on _Mx == 0 || _Cx <= _Mx, N4741 [rand.eng.lcong]/3
+    } else if _CONSTEXPR_IF (_Mx == 0) { // N4762 [rand.eng.lcong]/2: "If the template parameter m is 0, the modulus m
+                                         // used throughout this subclause [rand.eng.lcong] is
+                                         // numeric_limits<result_type>::max() plus 1." That is: Just do the multiply
+                                         // and let normal unsigned modulo take care of it
+        return static_cast<_Uint>(static_cast<_Uint>(_Ax * _Prev) + _Cx);
+    } else if _CONSTEXPR_IF (_Cx <= UINT_MAX
+                             && static_cast<_Uint>(_Mx - 1)
+                                    <= (UINT_MAX - _Cx)
+                                           / _Ax) { // unsigned int is sufficient to store intermediate calculation
```
Comment (StephanTLavavej) [inline stl/inc/random]: Now that `_CONSTEXPR_IF` isn't horribly damaging this formatting, it appears that this comment should be detached to improve readability.

File: stl/inc/random [hunk 3]
```diff
@@ -311,36 +311,28 @@ _NODISCARD _Uint _Next_linear_congruential_value(_Uint _Prev) noexcept {
...
+        const auto _Mul =
+            static_cast<unsigned int>(_Prev) * static_cast<unsigned int>(_Ax) + static_cast<unsigned int>(_Cx);
+        return static_cast<_Uint>(_Mul % _Mx);
+    } else if _CONSTEXPR_IF (_Cx <= ULLONG_MAX
+                             && static_cast<_Uint>(_Mx - 1)
+                                    <= (ULLONG_MAX - _Cx) / _Ax) { // unsigned long long is sufficient to store
+                                                                   // intermediate calculation
```
Comment (StephanTLavavej) [inline stl/inc/random]: Also detach comment?

File: stl/inc/system_error [hunk 4]
```diff
@@ -325,8 +325,8 @@ _NODISCARD inline bool operator<(const error_condition& _Left, const error_condi
 #endif // _STL_OPTIMIZE_SYSTEM_ERROR_OPERATORS
 
 // VIRTUALS FOR error_category
-_NODISCARD inline error_condition error_category::default_error_condition(int _Errval) const
-    noexcept { // make error_condition for error code
+_NODISCARD inline error_condition error_category::default_error_condition(
+    int _Errval) const noexcept { // make error_condition for error code
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: This comment should be detached according to our modern conventions.

File: stl/inc/system_error [hunk 5]
```diff
@@ -339,8 +339,8 @@ _NODISCARD inline bool error_category::equivalent(const error_code& _Code, int _
 }
 
 // MEMBER FUNCTIONS for error_code
-_NODISCARD inline error_condition error_code::default_error_condition() const
-    noexcept { // make error_condition for error code
+_NODISCARD inline error_condition
+    error_code::default_error_condition() const noexcept { // make error_condition for error code
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: Also detach.

File: stl/inc/type_traits [hunk 6]
```diff
@@ -2030,8 +2030,8 @@ struct _Is_swappable : _Is_swappable_with<add_lvalue_reference_t<_Ty>, add_lvalu
 
 // STRUCT TEMPLATE _Swap_cannot_throw
 template <class _Ty1, class _Ty2>
-struct _Swap_cannot_throw : bool_constant<noexcept(swap(_STD declval<_Ty1>(), _STD declval<_Ty2>()))
-                                          && noexcept(swap(_STD declval<_Ty2>(), _STD declval<_Ty1>()))> {
+struct _Swap_cannot_throw : bool_constant<noexcept(swap(_STD declval<_Ty1>(), _STD declval<_Ty2>()))&& noexcept(
```
Comment (StephanTLavavej) [inline stl/inc/type_traits]: Ugh, this is a regression in appearance. Add an empty comment to force wrapping before the logical AND?

File: stl/inc/variant [hunk 7]
```diff
@@ -876,8 +876,8 @@ public:
     _NODISCARD constexpr bool valueless_by_exception() const noexcept { // does this variant NOT hold a value?
         return _Which < 0;
     }
-    _NODISCARD constexpr size_t index() const
-        noexcept { // index of the contained alternative or variant_npos if valueless_by_exception
+    _NODISCARD constexpr size_t
+        index() const noexcept { // index of the contained alternative or variant_npos if valueless_by_exception
```
Comment (StephanTLavavej) [inline stl/inc/variant]: Detach comment.

**Comment by BillyONeal** [comment]
> So for me I assume that these are exclusively formatting changes in the other files and all functional changes between clang-9 / clang-10 will only come at a later stage?

Correct, except I'm going to have to make 2 functional changes to tests to make them pass.

File: stl/inc/functional (hunk 1 above)
Comment (BillyONeal) [reply stl/inc/functional]: I admittedly didn't even read what clang-format did :P

Comment (StephanTLavavej) [reply stl/inc/functional]: That's what I'm for, to read the output of humans and robots alike, line by repetitive line, until I find the single wrong one.

**Comment by CaseyCarter** [comment]
> Correct, except I'm going to have to make 2 functional changes to tests to make them pass.

Fixes for the feature-test macro test are [here](https://github.com/microsoft/STL/compare/master...CaseyCarter:allow_clang_10). I'm not sure what's going on with the spaceship test; it looks like cl runs are "failing" with exit status 0?

**Comment by cbezault** [comment]
They're XPASSing. It looks like there's improvement needed in the errors I output from the test harness in that case.

**Comment by CaseyCarter** [comment]
> They're XPASSing. It looks like there's improvement needed in the errors I output from the test harness in that case.

Ah: `C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28720\bin\HostX64\x86\cl.EXE` is Preview 2 - no wonder.

File: tests/std/expected_results.txt [hunk 8]
```diff
@@ -1,46 +1,2 @@
 # Copyright (c) Microsoft Corporation.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-
-# *** INSUFFICIENTLY NEW PUBLIC TOOLSET ***
-# TRANSITION, VS 2019 16.6p2
```
Comment (StephanTLavavej) [inline tests/std/expected_results.txt]: I believe you should update the README to mention 16.6 Preview 2 in both locations (currently it mentions Preview 1, so if someone follows that and then runs `tests/std`, these `expected_results` will be incorrect for them).

Comment (BillyONeal) [reply tests/std/expected_results.txt]: Heh, the downside of our script not mentioning this is I didn't even know it happened. Will do.

Comment (cbezault) [reply tests/std/expected_results.txt]: oh, sorry, just did it real quick, wanted to get the green checkmark asap.

File: tests/std/tests/P0513R0_poisoning_the_hash/test.cpp [hunk 9]
```diff
@@ -55,7 +55,7 @@ constexpr bool standard_hash_enabled() {
            && is_trivial_v<hash<T>> // as a consequence of the above
            && is_same_v<typename hash<T>::argument_type, T> //
            && is_same_v<typename hash<T>::result_type, size_t> //
-           && noexcept(hash<T>{}(declval<const T&>())) == NoExcept //
+                      && noexcept(hash<T>{}(declval<const T&>())) == NoExcept //
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply tests/std/tests/P0513R0_poisoning_the_hash/test.cpp]: I put an extra ()s around the noexcept and that seemed to fix it.

File: tests/std/expected_results.txt (hunk 8 above)
Comment (BillyONeal) [reply tests/std/expected_results.txt]: Nothing to be sorry about! Getting to work together like this is a great thing about Git

File: CMakeLists.txt [hunk 10]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) Microsoft Corporation.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-cmake_minimum_required(VERSION 3.15)
+cmake_minimum_required(VERSION 3.17)
```
Comment (cbezault) [inline CMakeLists.txt]: I don't think we can do this without also changing the instructions for building from the IDE in the README. Visual Studio 2019 16.6p2 currently ships with CMake 3.16.5 and so those build instructions would be broken by this change.

Comment (BillyONeal) [reply CMakeLists.txt]: Good point. I'll roll back the cmake edits.

Comment (BillyONeal) [reply CMakeLists.txt]: I added "Ensure dependent tools like CMake and LLVM match those shipping with the equivalently supported Visual Studio release." to https://github.com/microsoft/STL/wiki/Checklist-for-Updating-Toolset-and-or-Dependencies

File: README.md [hunk 11]
```diff
@@ -159,7 +159,7 @@ acquire this dependency.
 These instructions assume you're targeting `x64-windows`; you can change this constant below to target other
 architectures.
 
-1. Install [CMake][] 3.16.5 or later, [Ninja][] 1.10.0 or later, and Visual Studio 2019 16.6 Preview 1 or later.
+1. Install [CMake][] 3.17.0 or later, [Ninja][] 1.10.0 or later, and Visual Studio 2019 16.6 Preview 2 or later.
```
Comment (cbezault) [inline README.md]: This needs to say CMake 3.16.5

File: azure-devops/create-new-agent-image.ps1 [hunk 12]
```diff
@@ -8,7 +8,7 @@
 
 $Location = 'westus2'
 $Prefix = 'StlBuild-' + (Get-Date -Format 'yyyy-MM-dd')
-$VMSize = 'Standard_D16s_v3'
+$VMSize = 'Standard_F16s_v2'
```
Comment (cbezault) [inline azure-devops/create-new-agent-image.ps1]:
```suggestion
$VMSize = 'Standard_F32s_v2'
```

**Comment by StephanTLavavej** [comment]
I observe that it's difficult to review changes when you squash away history :crying_cat_face:

**Comment by StephanTLavavej** [comment]
Fortunately everything looks good (GitHub at least remembers which files have been Viewed and are unchanged). Thanks.

**Comment by BillyONeal** [comment]
> I observe that it's difficult to review changes when you squash away history 😿

Sorry, without squashing it was getting awkward to apply to the msvc repo :/

**Comment by StephanTLavavej** [comment]
I'm not sure how you port changes from GitHub to MSVC, but I've found that I can port arbitrarily complex histories to MSVC. As long as the GitHub branch is either rebased *or merged* with the latest `master`, a simple `git diff master WHATEVER_BRANCH > meow.patch` is sufficient to create an inherently squashed patch to apply to MSVC.

**Pull request merged**